File: frontend/src/utils/planLimits.ts
/**
 * Plan limits utilities for frontend feature gating.
 * 
 * Mirrors backend/app/core/plan_limits.py for consistent enforcement.
 */

import type { UserPlan } from '../context/AuthContext';

// =============================================================================
// Sport IDs (matching backend)
// =============================================================================

export const SPORT_NBA = 30;
export const SPORT_NFL = 31;
export const SPORT_NCAAB = 32;
export const SPORT_WNBA = 34;
export const SPORT_MLB = 40;
export const SPORT_NCAAF = 41;
export const SPORT_ATP = 42;
export const SPORT_WTA = 43;
export const SPORT_NHL = 53;
export const SPORT_PGA = 60;
export const SPORT_EPL = 70;
export const SPORT_UCL = 71;
export const SPORT_MLS = 72;
export const SPORT_UEL = 73;
export const SPORT_UECL = 74;
export const SPORT_UFC = 80;

// Sports available on free tier
export const FREE_SPORTS = [SPORT_NBA, SPORT_NFL];

// All available sports
export const ALL_SPORTS = [
  SPORT_NBA, SPORT_NFL, SPORT_NCAAB, SPORT_WNBA, SPORT_MLB, SPORT_NCAAF,
  SPORT_ATP, SPORT_WTA, SPORT_NHL, SPORT_PGA,
  SPORT_EPL, SPORT_UCL, SPORT_MLS, SPORT_UEL, SPORT_UECL,
  SPORT_UFC,
];

// =============================================================================
// Feature Types
// =============================================================================

export type PlanFeature =
  | 'live_ev'
  | 'alt_lines'
  | 'watchlists'
  | 'backtest'
  | 'analytics'
  | 'model_performance'
  | 'my_edge'
  | 'stats_filters'
  | 'export_data'
  | 'unlimited_props'
  | 'unlimited_parlay_legs';

// =============================================================================
// Plan Limits Configuration
// =============================================================================

interface PlanLimits {
  allowedSports: number[] | 'all';
  dailyPropsLimit: number | null;  // null = unlimited
  parlayMaxLegs: number | null;    // null = unlimited
  features: Record<PlanFeature, boolean>;
}

const PLAN_LIMITS: Record<UserPlan, PlanLimits> = {
  free: {
    allowedSports: FREE_SPORTS,
    dailyPropsLimit: 10,
    parlayMaxLegs: 3,
    features: {
      live_ev: false,
      alt_lines: false,
      watchlists: false,
      backtest: false,
      analytics: false,
      model_performance: false,
      my_edge: false,
      stats_filters: false,
      export_data: false,
      unlimited_props: false,
      unlimited_parlay_legs: false,
    },
  },
  trial: {
    allowedSports: 'all',
    dailyPropsLimit: null,
    parlayMaxLegs: null,
    features: {
      live_ev: true,
      alt_lines: true,
      watchlists: true,
      backtest: true,
      analytics: true,
      model_performance: true,
      my_edge: true,
      stats_filters: true,
      export_data: true,
      unlimited_props: true,
      unlimited_parlay_legs: true,
    },
  },
  pro: {
    allowedSports: 'all',
    dailyPropsLimit: null,
    parlayMaxLegs: null,
    features: {
      live_ev: true,
      alt_lines: true,
      watchlists: true,
      backtest: true,
      analytics: true,
      model_performance: true,
      my_edge: true,
      stats_filters: true,
      export_data: true,
      unlimited_props: true,
      unlimited_parlay_legs: true,
    },
  },
};

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Get plan limits for a user plan.
 */
export function getPlanLimits(plan: UserPlan): PlanLimits {
  return PLAN_LIMITS[plan] || PLAN_LIMITS.free;
}

/**
 * Get list of sport IDs the user can access.
 */
export function getAllowedSports(plan: UserPlan): number[] {
  const limits = getPlanLimits(plan);
  return limits.allowedSports === 'all' ? ALL_SPORTS : limits.allowedSports;
}

/**
 * Check if user can access a specific sport.
 */
export function canAccessSport(plan: UserPlan, sportId: number): boolean {
  const allowed = getAllowedSports(plan);
  return allowed.includes(sportId);
}

/**
 * Check if user can use a specific feature.
 */
export function canUseFeature(plan: UserPlan, feature: PlanFeature): boolean {
  const limits = getPlanLimits(plan);
  return limits.features[feature] || false;
}

/**
 * Get daily props limit for a plan.
 * Returns null for unlimited.
 */
export function getDailyPropsLimit(plan: UserPlan): number | null {
  const limits = getPlanLimits(plan);
  return limits.dailyPropsLimit;
}

/**
 * Get max parlay legs for a plan.
 * Returns null for unlimited.
 */
export function getMaxParlayLegs(plan: UserPlan): number | null {
  const limits = getPlanLimits(plan);
  return limits.parlayMaxLegs;
}

/**
 * Check if a user has reached their daily props limit.
 */
export function hasReachedPropsLimit(plan: UserPlan, propsViewed: number): boolean {
  const limit = getDailyPropsLimit(plan);
  if (limit === null) return false;
  return propsViewed >= limit;
}

/**
 * Check if a parlay exceeds the user's leg limit.
 */
export function exceedsParlayLimit(plan: UserPlan, legs: number): boolean {
  const limit = getMaxParlayLegs(plan);
  if (limit === null) return false;
  return legs > limit;
}

// =============================================================================
// Sport Name Helpers
// =============================================================================

export const SPORT_NAMES: Record<number, string> = {
  [SPORT_NBA]: 'NBA',
  [SPORT_NFL]: 'NFL',
  [SPORT_NCAAB]: 'NCAAB',
  [SPORT_WNBA]: 'WNBA',
  [SPORT_MLB]: 'MLB',
  [SPORT_NCAAF]: 'NCAAF',
  [SPORT_ATP]: 'ATP Tennis',
  [SPORT_WTA]: 'WTA Tennis',
  [SPORT_NHL]: 'NHL',
  [SPORT_PGA]: 'PGA Tour',
  [SPORT_EPL]: 'EPL',
  [SPORT_UCL]: 'Champions League',
  [SPORT_MLS]: 'MLS',
  [SPORT_UEL]: 'Europa League',
  [SPORT_UECL]: 'Conference League',
  [SPORT_UFC]: 'UFC',
};

export function getSportName(sportId: number): string {
  return SPORT_NAMES[sportId] || `Sport ${sportId}`;
}

/**
 * Get locked sports (not available on free plan).
 */
export function getLockedSports(): number[] {
  return ALL_SPORTS.filter(id => !FREE_SPORTS.includes(id));
}

/**
 * Check if a sport is locked for free users.
 */
export function isSportLocked(sportId: number): boolean {
  return !FREE_SPORTS.includes(sportId);
}

File: backend/app/core/plan_limits.py
"""Plan-based feature limits configuration."""

from dataclasses import dataclass
from typing import Optional

from app.models.user import User, UserPlan


# =============================================================================
# Plan Limits Configuration
# =============================================================================

@dataclass
class PlanLimits:
    """Feature limits for a subscription plan."""
    
    # Sports access
    allowed_sports: list[int] | str  # List of sport IDs or "all"
    
    # Props limits
    daily_props_limit: Optional[int]  # None = unlimited
    
    # Features
    stats_filters: bool  # Can use market/side filters in Stats
    stats_hot_cold: bool  # Can see Hot/Cold players
    stats_streaks: bool  # Can see streaks
    parlay_max_legs: Optional[int]  # None = unlimited
    
    # Premium features
    live_ev: bool
    alt_lines: bool
    watchlists: bool
    backtest: bool
    analytics: bool
    model_performance: bool
    my_edge: bool
    export_data: bool


# Sport IDs
SPORT_NBA = 30
SPORT_NFL = 31
SPORT_NCAAB = 32
SPORT_MLB = 40
SPORT_NCAAF = 41
SPORT_ATP = 42
SPORT_WTA = 43
SPORT_NHL = 53

FREE_SPORTS = [SPORT_NBA, SPORT_NFL]
ALL_SPORTS = [SPORT_NBA, SPORT_NFL, SPORT_NCAAB, SPORT_MLB, SPORT_NCAAF, SPORT_ATP, SPORT_WTA, SPORT_NHL]


# =============================================================================
# Plan Definitions
# =============================================================================

PLAN_LIMITS: dict[str, PlanLimits] = {
    UserPlan.FREE.value: PlanLimits(
        # Limited sports
        allowed_sports=FREE_SPORTS,
        
        # Limited props per day
        daily_props_limit=10,
        
        # Basic stats only
        stats_filters=False,
        stats_hot_cold=True,  # Can see summary
        stats_streaks=True,
        parlay_max_legs=3,
        
        # No premium features
        live_ev=False,
        alt_lines=False,
        watchlists=False,
        backtest=False,
        analytics=False,
        model_performance=False,
        my_edge=False,
        export_data=False,
    ),
    
    UserPlan.TRIAL.value: PlanLimits(
        # Full sports access
        allowed_sports="all",
        
        # Unlimited props
        daily_props_limit=None,
        
        # Full stats
        stats_filters=True,
        stats_hot_cold=True,
        stats_streaks=True,
        parlay_max_legs=None,
        
        # All premium features
        live_ev=True,
        alt_lines=True,
        watchlists=True,
        backtest=True,
        analytics=True,
        model_performance=True,
        my_edge=True,
        export_data=True,
    ),
    
    UserPlan.PRO.value: PlanLimits(
        # Full sports access
        allowed_sports="all",
        
        # Unlimited props
        daily_props_limit=None,
        
        # Full stats
        stats_filters=True,
        stats_hot_cold=True,
        stats_streaks=True,
        parlay_max_legs=None,
        
        # All premium features
        live_ev=True,
        alt_lines=True,
        watchlists=True,
        backtest=True,
        analytics=True,
        model_performance=True,
        my_edge=True,
        export_data=True,
    ),
}


# =============================================================================
# Helper Functions
# =============================================================================

def get_plan_limits(user: Optional[User]) -> PlanLimits:
    """
    Get plan limits for a user.
    
    Anonymous users get free tier limits.
    """
    if user is None:
        return PLAN_LIMITS[UserPlan.FREE.value]
    
    # Active trial counts as pro
    if user.is_trial:
        return PLAN_LIMITS[UserPlan.PRO.value]
    
    return PLAN_LIMITS.get(user.plan, PLAN_LIMITS[UserPlan.FREE.value])


def get_allowed_sports(user: Optional[User]) -> list[int]:
    """Get list of sport IDs the user can access."""
    limits = get_plan_limits(user)
    
    if limits.allowed_sports == "all":
        return ALL_SPORTS
    
    return limits.allowed_sports


def can_access_sport(user: Optional[User], sport_id: int) -> bool:
    """Check if user can access a specific sport."""
    allowed = get_allowed_sports(user)
    return sport_id in allowed


def check_daily_props_limit(user: Optional[User]) -> tuple[bool, int, Optional[int]]:
    """
    Check if user has hit their daily props limit.
    
    Returns:
        (can_view_more, viewed_today, limit)
    """
    limits = get_plan_limits(user)
    
    if limits.daily_props_limit is None:
        return (True, 0, None)
    
    if user is None:
        # Anonymous users: simple limit
        return (True, 0, limits.daily_props_limit)
    
    viewed = user.props_viewed_today or 0
    can_view = viewed < limits.daily_props_limit
    
    return (can_view, viewed, limits.daily_props_limit)


def can_use_feature(user: Optional[User], feature: str) -> bool:
    """
    Check if user can use a specific feature.
    
    Features: live_ev, alt_lines, watchlists, backtest, analytics,
              model_performance, my_edge, export_data, stats_filters
    """
    limits = get_plan_limits(user)
    return getattr(limits, feature, False)


# =============================================================================
# Response Helpers
# =============================================================================

def get_plan_info_response(user: Optional[User]) -> dict:
    """
    Get plan info for API response.
    
    Useful for frontend to know what features are available.
    """
    limits = get_plan_limits(user)
    can_view, viewed, limit = check_daily_props_limit(user)
    
    return {
        "plan": user.plan if user else "free",
        "is_pro": user.is_pro if user else False,
        "is_trial": user.is_trial if user else False,
        "trial_days_left": user.trial_days_left if user else None,
        "allowed_sports": get_allowed_sports(user),
        "daily_props_limit": limit,
        "props_viewed_today": viewed,
        "can_view_more_props": can_view,
        "features": {
            "live_ev": limits.live_ev,
            "alt_lines": limits.alt_lines,
            "watchlists": limits.watchlists,
            "backtest": limits.backtest,
            "analytics": limits.analytics,
            "model_performance": limits.model_performance,
            "my_edge": limits.my_edge,
            "stats_filters": limits.stats_filters,
            "export_data": limits.export_data,
        },
    }
